### ISSUE TITLE
[SUGGESTION] automatically ensure that copy-on-write containers do not unintentionally detach

### ISSUE THREAD

**Issue by vladimir-kraus** [hsutter/cppfront#226]
This may be more a question rather than a suggestion. Maybe this problem has already been assessed and tackled...

Problem explanation: Some containers are copy-on-write (COW). They are especially popular for example in Qt. COW containers provide many benefits but also some dangers. One of the dangers affecting performance is that when used incorrectly COW containers make copy of themselves (aka they "detach"). This basically happens when there are more than one reference to the underlying data AND any non-const method is called for the container. In that case the container must assume its content is going to be changed and it makes a copy of itself to ensure that other references point to the original unchanged data. Very often this is not something that we want to, it often is a result of programming mistake. And this is also a problem in range for loops since they are calling begin() and end() which use their non-const variants when called on non-const container.

It would really nice feature of cppfront if this kind of bugs was taken care of automatically by cppfront during transpilation to cpp. Though I am not yet acquainted with cppfront enough to know if this is already implemented, or planned or whether it is even possible. Cppfront would have to somehow know if it can use const or non-const methods. I have seen there are some out/in/inout modifiers, probably this is the way it could know whether to use const or non-const code.

PS: In any case, congratulations to absolutely amazing project and big thanks to anyone involved in it, especially to Herb. As a Qt programmer I am looking forward to using Qt with cppfront. :)

**Comment by hsutter** [comment]
Thanks for the kind words and encouragement, I appreciate it.

Can you give more examples of the problem and current workarounds?

> This basically happens when there are more than one reference to the underlying data AND any non-const method is called for the container. [...] it often is a result of programming mistake, e.g. using begin() or end() instead of cbegin() or cend(). And this is also a problem in range for loops since they are calling begin() and end().

So I'm imagining code something like this (https://godbolt.org/z/E6Kb3sPrh):

```
auto myfunc (CowContainer& container) {
    for( auto& elem : container ) {  // calls .begin() /*non-const*/
        std::cout << elem;
    }
}
```

Is that the kind of problem you're seeing?

What if you take the parameter by `const&` like this:

```
auto myfunc (CowContainer const& container) {
    for( auto& elem : container ) {  // calls .begin() /*non-const*/
        std::cout << elem;
    }
}
```

Is that the kind of thing that helps your problem, and you'd like to make it easier to pass things as `const` and not forget to do that?

If so, I think Cpp2 helps you a lot because Cpp2 function parameters are `const` by default. For example, writing this in Cpp2:

```
myfunc: (container: CowContainer) = {
    for container do :(elem:_) = {
        std::cout << elem;
    }
}
```

is the same as this in Cpp1 (in fact this is the generated code, I just cut-and-pasted it)... here `cpp2::in<CowContainer>` is an alias for `CowContainer const&`, plus as a bonus the `elem` parameter to the loop body is also `const` by default (https://godbolt.org/z/xoo68PWxP):

```
auto myfunc(cpp2::in<CowContainer> container) -> void{
    for ( auto&& cpp2_range = container;  auto const& elem : cpp2_range ) {
        std::cout << elem;
    }
}
```

Does that help answer your question?

**Comment by hsutter** [comment]
Aside: writing this example

```
    for container do :(elem:_) = {
        std::cout << elem;
    }
```

finally pushed me over the edge to do an item on my to-do list, which is to default function parameters to be of deduced type without having to write `: _` explicitly... so with commit 7a842a67569bf288ffd85d5f1b204e498683d721, this now works and should help reduce the "line noise" perception of unnamed (lambda) functions:

```
    for container do :(elem) = {
        std::cout << elem;
    }
```

Here `elem` is still `in` (and therefore `const`) by default.

**Comment by vladimir-kraus** [comment]
I think in practice the main problem is not then the container is passed as function parameter because hopefully all C++ programmers are used to pass input parameters of non trivial types by const reference. So the problematic use cases are basically of the following two types. (I will be using {QVector} example as this is typical COW container used in Qt)

Usecase 1 - having 2 (or more) containers referencing the same underlying data:
```
QVector v1 = {1, 2, 3}; // create COW container with refcount == 1
...
auto v2 = v1; // we make a copy which increases refcount to 2
...
for( const auto& elem : v1 ) {  // calls .begin() /*non-const*/ and therefore makes a copy of v1 data (so called detachment)
   std::cout << elem;
}
```

Usecase 2 - we iterate on a container returned from function:
```
QVector<int> createVector();
...
for( const auto& elem : createVector() ) {  // calls .begin() /*non-const*/ but we do not know the refcount (so detachment may happen)
   std::cout << elem;
}
```

In this case the returned container may or may not have a ref count bigger than 1. Nobody can know without seeing what is inside function createVector() because if it returns a copy of already existing vector, then the ref count will be bigger than 1 and then the detachment will happen.

The problem is typically solved in Qt by using {std::as_const()} or Qt alternative {qAsConst()}:
```
...
for( const auto& elem : std::as_const(v1) ) { // it forces to call const .begin()
...
```

Another alternative is to take a const reference of the container before the use:
```
...
const auto &vref = v1;
for( const auto& elem : vref  ) { // calls .begin() /*const*/
...
```

But in practice this is often forgotten especially by less skilled programmers (such as myself :) ). And then it may lead to detachment (i.e. a copy of the whole container) which is subsequently deleted again. The program works well without any error. But the performance suffers.

My question is whether it would somehow be possible to automatically iterate not on the container but on a const reference to it (or use {as_const}) before the container is used in a loop iif its elements are accessed as {in} parameter, i.e. they are read-only. Honestly I am not familiar with cppfront or able to foresee all possible consequences it might have...

**Comment by switch-blade-stuff** [comment]
I assume the emitted code can instead use `std::ranges::for_each`, then try to check if it can be invoked using const container with something like `if constexpr (requires { std::ranges::for_each(std::as_const(range), body); })` and invoke it as such if possible.

This does, however, require the use of the `for_each` algorithm instead of the builtin range-for loop and will introduce more boilerplate into the emitted code.

It does also have the potential advantages of enforcing const-correctness though.

**Comment by JohelEGP** [comment]
`for_each` is not a substitute for the `for` statement. See https://github.com/ericniebler/range-v3/issues/1372.

**Comment by vladimir-kraus** [comment]
Since I noticed label "further information requested" maybe I should summarise the suggestion:

When a user iterates over a collection in for-loop AND the iterated element type is read-only (`in`) then the collection should be casted to be const with {std::as_const} or some other mechanism (e.g. taking a const reference to it and iterating over it) should be applied to force only const calling const version of `begin()` and `end()` in the for-loop. This will avoid the unwanted situation when a copy-on-write container will detach. 

As I wrote above I am not able to foresee all possible consequences of such a change and therefore I do not know if it is possible.

**Comment by hsutter** [comment]
tl;dr: OK, you've convinced me. Thanks! Implemented: de630b93dbe7b715328a6ad67e1263d16e691e86

@vladimir-kraus A request please: If you are able to, can you try the updated cppfront with `QVector` and verify that it now does correctly prevent a COW operation in a `for` loop using the default `in` parameter?

---

Longer answer... What you wrote is a good summary:

> When a user iterates over a collection in for-loop AND the iterated element type is read-only (`in`) then the collection should be casted to be const with {std::as_const} or some other mechanism (e.g. taking a const reference to it and iterating over it) should be applied to force only const calling const version of `begin()` and `end()` in the for-loop.

What convinces me is that this is directly aligned with my indent for loops with `in` loop elements to declaratively be read-only loops. See [d0708 Parameter passing, section 1.4 (page 19)](https://github.com/hsutter/708/blob/main/708.pdf). So I agree that declared intent should be propagated when we bind the range variable too -- that's a great suggestion for improving the implementation, thanks. This should give you specific benefit you're looking for, and other benefits from propagating the constness we know about so that the second (Cpp1, C++) compiler and libraries can take advantage of it.

**Comment by vladimir-kraus** [comment]
I thank you. I think that this feature will make cppfront very popular among the whole Qt community... I think that all of them are very tired of constant thinking about unwanted detachment whenever they use range for loop. Of course, there is still a danger of unwanted detachment whenever they use any non-const method but there is always a const-only alternative (e.g. `QVector::constFirts()` instead of `QVector::first()`) so this is up to the programmer to choose the right one and up to static analyzer to warn him if he chooses the wrong one. But for range loops this constness-enforcing is a great help.
